### PR TITLE
Implement custom reporters

### PR DESCRIPTION
**Pull request by dekrain**

### CODE REVIEW

**Comment by dekrain** [comment]
This not work very good. Try to put `if` block to definition and it break

**Comment by TheBrokenRail** [comment]
Try Using startHats(() and copying code from "broadcast and wait" block

**Comment by TheBrokenRail** [comment]
I made PR For startHats

**Comment by tmickel** [comment]
Hey @dekrain and @TheBrokenRail! Cool to be thinking about this problem.

I'm worried about storing values on the `Scratch3ProcedureBlocks` object. What happens when you run two custom reporters in parallel? I could imagine that if both report at the same time, something weird would happen.

Traditionally programming languages implement this sort of thing (returning values from a procedure) using a stack:
https://en.wikipedia.org/wiki/Call_stack

We have most of the infrastructure needed in `thread.stack` and `thread.stackFrame` - the former which tracks descent into the block stack, and the latter of which is intended for storing arbitrary data, like parameters for custom procedures and (potentially) reported values.

A correct implementation of custom blocks and custom reporters will need to use a stack associated with the thread in some form or another. Otherwise, crazy things seem likely to happen when two blocks execute in parallel or one thread yields for a frame.

**Comment by tmickel** [comment]
P.s., I checked out this PR and tried to run it, but I get this error:
`vm.js:5206 Uncaught TypeError: Cannot read property 'length' of undefined`

**Comment by dekrain** [comment]
Ok, I've read scratch-vm code many times and I have not see any problems. I'm annoyed.

File: src/blocks/scratch3_procedures.js [hunk 1]
```diff
@@ -4,6 +4,8 @@ function Scratch3ProcedureBlocks(runtime) {
      * @type {Runtime}
      */
     this.runtime = runtime;
+    this.REPORT = {};
+    this.currentCall = null;
```
Comment (tmickel) [inline src/blocks/scratch3_procedures.js]: We can't store data on the `Scratch3ProcedureBlocks` object, as I previously mentioned. Any data that's relevant for procedure calls needs to be associated with the thread.

File: src/blocks/scratch3_procedures.js [hunk 2]
```diff
@@ -13,14 +15,34 @@ function Scratch3ProcedureBlocks(runtime) {
...
+    };
+};
+
+Scratch3ProcedureBlocks.prototype.getHats = function() {
+    return {
+        'procedures_defnoreturn': {
+            restartExistingThreads: false
+        },
+        'procedures_defreturn': {
+            restartExistingThreads: false
```
Comment (tmickel) [inline src/blocks/scratch3_procedures.js:31]: These aren't hat blocks, and the mechanism for starting hats isn't intended for starting procedures.

File: src/blocks/scratch3_procedures.js [hunk 3]
```diff
@@ -13,14 +15,34 @@ function Scratch3ProcedureBlocks(runtime) {
...
     };
 };
 
 Scratch3ProcedureBlocks.prototype.defNoReturn = function () {
     // No-op: execute the blocks.
 };
 
+Scratch3ProcedureBlocks.prototype.defReturn = function (args) {
+    var name = args.NAME;
+    this.REPORT[name] = ((typeof this.REPORT[name] === 'undefined') ?
```
Comment (tmickel) [inline src/blocks/scratch3_procedures.js]: What if you run two of the same custom reporter in parallel?

File: src/blocks/scratch3_procedures.js [hunk 4]
```diff
@@ -29,4 +51,48 @@ Scratch3ProcedureBlocks.prototype.callNoReturn = function (args, util) {
...
+            delete this.REPORT[procedureName];
+            this.currentCall = this.currentCall.parent;
+            return rep;
+        }
+    }
+};
+
+Scratch3ProcedureBlocks.prototype.report = function (args) {
+    if (this.currentCall) {
+        this.REPORT[this.currentCall.name] = args.VALUE;
```
Comment (tmickel) [inline src/blocks/scratch3_procedures.js]: Same comment as above

**Comment by tmickel** [comment]
@dekrain I'm not sure how fruitful it's going to be to work on this without digging into the details of the VM. The Scratch team was planning to do an implementation of this sometime in the next several weeks/months.

**Comment by thisandagain** [comment]
@dekrain Thanks for taking a first crack at this. 

For big feature implementations like this we would prefer to talk about the approach before generating a bunch of code and a PR. This helps reduce the amount of work for you (and for us) and makes sure that we are all on the same page before charging into a solution. As @tmickel mentioned we are looking to produce an implementation for this over the coming weeks/months, but in the meantime I think it would be most fruitful to continue this discussion here:
https://github.com/LLK/scratch-vm/issues/79

**Pull request closed without merging**